test/jdk/javax/swing/text/html/HTMLDocument/HTMLTextDecoration.java

```diff
@@ -0,0 +1,157 @@
+            <head>
+                <meta charset="UTF-8">
+                <title>underline + line-through text</title>
+                <style>
+                    .underline   { text-decoration: underline }
+                    .lineThrough { text-decoration: line-through }
+                </style>
+            </head>
+            <body>
+            <p><u><span style='text-decoration: line-through'>underline + line-through?</span></u></p>
+            <p><s><span style='text-decoration: underline'>underline + line-through?</span></s></p>
+            <p><strike><span style='text-decoration: underline'>underline + line-through?</span></strike></p>
+
+            <p><span style='text-decoration: line-through'><u>underline + line-through?</u></span></p>
+            <p><span style='text-decoration: underline'><s>underline + line-through?</s></span></p>
+            <p><span style='text-decoration: underline'><strike>underline + line-through?</strike></span></p>
+
+            <p><span style='text-decoration: line-through'><span style='text-decoration: underline'>underline + line-through?</span></span></p>
+            <p><span style='text-decoration: underline'><span style='text-decoration: line-through'>underline + line-through?</span></span></p>
+
+            <p style='text-decoration: line-through'><u>underline + line-through?</u></p>
+            <p style='text-decoration: underline'><s>underline + line-through?</s></p>
+            <p style='text-decoration: underline'><strike>underline + line-through?</strike></p>
+
+            <p style='text-decoration: line-through'><span style='text-decoration: underline'>underline + line-through?</span></p>
+            <p style='text-decoration: underline'><span style='text-decoration: line-through'>underline + line-through?</span></p>
+
+            <p class="underline"><span class="lineThrough">underline + line-through?</span></p>
+            <p class="underline"><s>underline + line-through?</s></p>
+            <p class="underline"><strike>underline + line-through?</strike></p>
+
+            <p class="lineThrough"><span class="underline">underline + line-through?</span></p>
+            <p class="lineThrough"><u>underline + line-through?</u></p>
+
+            <div class="underline"><span class="lineThrough">underline + line-through?</span></div>
+            <div class="underline"><s>underline + line-through?</s></div>
+            <div class="underline"><strike>underline + line-through?</strike></div>
+
+            <div class="lineThrough"><span class="underline">underline + line-through?</span></div>
+            <div class="lineThrough"><u>underline + line-through?</u></div>
+
+            <div class="underline"><p class="lineThrough">underline + line-through?</p></div>
+            <div class="lineThrough"><p class="underline">underline + line-through?</p></div>
+
+            <div class="underline"><div class="lineThrough">underline + line-through?</div></div>
```

Comment:
The test fails when multiple CSS classes are on the same tag. Are we considering such cases as part of this fix?

For instance `<div class="underline lineThrough">This text should be underlined</div>`
In this case, the 1st class - underline is applied and lineThrough is ignored since both are related to text-decoration when tested on a HTML editor.
But when tested using HTMLTestDecoration.java it is not underlined as expected.
For example: 
```
<!DOCTYPE html>
<html lang="en">
   <head>
      <meta charset="UTF-8">
      <title>underline + line-through text</title>
      <style>
         .underline   { text-decoration: underline }
         .lineThrough { text-decoration: line-through }
      </style>
   </head>
   <body>
      <div class="underline lineThrough">This text should be underlined</div>
   </body>
</html>

```

Reply:
I'm pretty sure there are other cases where `text-decoration` isn't resolved correctly. It is not a goal to resolve all of the possible issues.

I'll check whether this scenario can be handled easily. I don't think it is. If not, I'll submit another bug.

Reply:
Sounds good.

Reply:
Swing does not support *multiple classes* on an HTML element.

I submitted [JDK-8330119](https://bugs.openjdk.org/browse/JDK-8330119): _No support for multiple HTML classes in Swing_. It may never be fixed…